Add updatepeer RPC

jnewbery opened the pull request:
Adds an `updatepeer` RPC to update peer config and carry out actions on the peer by NodeId. At the moment, the actions are just changing whether the peer is whitelisted and whether the peer is a manual_connection (previously called an addnode). Future possible actions would be changing banscore, banning, and so on.

This is designed to be called using named arguments, but due to the RPC infrastructure can also be called with positional arguments (although doing so would be very fiddly).

I've set the category to hidden for now. We may want to make the whitelisting behaviour more granular in future and I don't want to commit us to a public API that we can't then change.

jnewbery committed:
[RPC] Refactor getpeerinfo RPC to call helper function NodeStatsToJSON().

This commit refactors getpeerinfo() to call a helper function
NodeStatsToJSON() to build a JSON object with peer information.

jnewbery committed:
[RPC] Add GetNodeStats() function for individual node.

GetNodeStats() can now be called for a single node.

jnewbery committed:
[RPC] Allow getpeerinfo to be called for a single node.

The getpeerinfo RPC can now be called for a single node by specifying the
node id.

jnewbery committed:
[RPC] Add updatepeer RPC.

This commit adds an updatepeer RPC for updating a single peer node.

Currently, updatepeer can be used to update the whitelisting and
manual_connection settings for the node. It could potentially be
extended in future to allow updating other attributes of the peer node.

theuni commented:
I'm not sure I like the idea of adding a command that changes the implied state of a peer via its attributes, rather than performing actions explicitly. The whitelist side-effects are especially hazy, see the brief discussion here: https://github.com/bitcoin/bitcoin/pull/10051#issuecomment-290622674.

jonasschnelli commented:
@jnewbery:
Can you elaborate the use-case for `disconnect`? IMO `setban` provides a similar interface, with disconnecting & banning for a specific timespan (`1h`, etc.) because, a pure `disconnect` does not prevent the peer from a direct re-connect.

jonasschnelli commented:
Ah... an there is already `disconnectnode`.

jnewbery commented:
@jonasschnelli see #2729 and #6271 for history of the `disconnectnode` RPC. It's also useful in testing to be able to control the topology of the test nodes.

@theuni - I agree that whitelisting is a mess and should be broken out into bits for controlling individual behaviours. Hence my original comment: "I have a feeling we may want to make the whitelisting behaviour more granular and I don't want to commit us to a public API that we can't then change." I was thinking of your comment in #10051 but couldn't find the reference.

> the idea of adding a command that changes the implied state of a peer via its attributes, rather than performing actions explicitly

I don't understand what this means in the context of whitelisting. The idea is to update the peer's whitelist behaviour. What would performing actions explicitly entail?

jnewbery commented:
Closing in favour of #10143 for the immediate need (disconnect node by id), but I think this could still be a useful RPC in the future.

jnewbery commented:
Reopening with just the ability to update:

- `fWhitelisted`
- `m_manual_connection`

The net_processing functionality for the v0.15.0.2 PRs is disabled for manual connections, so this PR could be helpful for testing those changes.

@sdaftuar

promag commented on src/rpc/net.cpp:
```diff
@@ -121,73 +182,79 @@ UniValue getpeerinfo(const JSONRPCRequest& request)
-        for (const mapMsgCmdSize::value_type &i : stats.mapRecvBytesPerMsgCmd) {
-            if (i.second > 0)
-                recvPerMsgCmd.push_back(Pair(i.first, i.second));
+    if (request.params.size() >= 3) {
```
Is null instead.

promag commented on src/rpc/net.cpp:
```diff
@@ -121,73 +182,79 @@ UniValue getpeerinfo(const JSONRPCRequest& request)
+    UniValue entry(UniValue::VOBJ);
+    CNodeStats stats;
+
+    if (g_connman->GetNodeStats(nodeid, stats)) {
```
If not then raise invalid parameter?

promag replied:
Missing test for invalid node.

promag commented on src/rpc/net.cpp:
```diff
@@ -121,73 +182,79 @@ UniValue getpeerinfo(const JSONRPCRequest& request)
     UniValue ret(UniValue::VARR);
 
-    for (const CNodeStats& stats : vstats) {
+    if (request.params.size() >= 1 && request.params[0].isNum()) {
```
Not null instead.

promag reviewed (commented):
Concept ACK. Will test.

NicolasDorier commented:
whitelisting via rpc?? BIG BIG concept ACK. It will make configuration of services depending on Bitcoin Core RPC and P2P so much easier.

promag commented on src/rpc/net.cpp:
```diff
@@ -121,73 +182,81 @@ UniValue getpeerinfo(const JSONRPCRequest& request)
+        CNodeStats stats;
+
+        int node_id = request.params[0].get_int();
+        if (g_connman->GetNodeStats(node_id, stats)) {
```
```cpp
if (!...) {
    throw ...;
}
```

promag commented on src/rpc/net.cpp:
```diff
@@ -65,12 +65,73 @@ UniValue ping(const JSONRPCRequest& request)
 UniValue getpeerinfo(const JSONRPCRequest& request)
 {
-    if (request.fHelp || request.params.size() != 0)
+    if (request.fHelp || request.params.size() >= 2)
```
Nit, `> 1`.

promag replied:
Needs release notes.

promag commented on src/rpc/net.cpp:
```diff
@@ -65,12 +65,73 @@ UniValue ping(const JSONRPCRequest& request)
+
+    UniValue sendPerMsgCmd(UniValue::VOBJ);
+    for (const mapMsgCmdSize::value_type &i : stats.mapSendBytesPerMsgCmd) {
+        if (i.second > 0)
```
Nit, same line or `{ }`.

promag commented on src/net.cpp:
```diff
@@ -2521,6 +2521,18 @@ size_t CConnman::GetNodeCount(NumConnections flags)
+bool CConnman::GetNodeStats(NodeId nodeid, CNodeStats& stats)
+{
+    LOCK(cs_vNodes);
+    for(CNode* pnode : vNodes) {
```
Nit, missing spaces after `for`.

promag commented on src/rpc/net.cpp:
```diff
@@ -65,12 +65,73 @@ UniValue ping(const JSONRPCRequest& request)
+
+    UniValue recvPerMsgCmd(UniValue::VOBJ);
+    for (const mapMsgCmdSize::value_type &i : stats.mapRecvBytesPerMsgCmd) {
+        if (i.second > 0)
```
Same as above.

promag commented on src/rpc/net.cpp:
```diff
@@ -121,73 +182,81 @@ UniValue getpeerinfo(const JSONRPCRequest& request)
+
+UniValue updatepeer(const JSONRPCRequest& request)
+{
+    if (request.fHelp || request.params.size() > 3 || request.params.size() == 0)
```
Nit, the most common expression is
```cpp
if (request.fHelp || request.params.size() < 1 || request.params.size() > 3)
```

promag commented on src/rpc/net.cpp:
```diff
@@ -121,73 +182,81 @@ UniValue getpeerinfo(const JSONRPCRequest& request)
+            + HelpExampleRpc("updatepeer", "0, true")
+        );
+
+    if(!g_connman) {
```
Nit, space after `if`.

promag commented on src/rpc/net.cpp:
```diff
@@ -121,73 +182,81 @@ UniValue getpeerinfo(const JSONRPCRequest& request)
+    LOCK(cs_main);
+
+    RPCTypeCheckArgument(request.params[0], UniValue::VNUM);
+    int nodeid = request.params[0].get_int();
```
Same validation of `getpeerinfo`.

promag commented on src/rpc/net.cpp:
```diff
@@ -121,73 +182,81 @@ UniValue getpeerinfo(const JSONRPCRequest& request)
-            obj.push_back(Pair("inflight", heights));
+        CNodeStats stats;
+
+        int node_id = request.params[0].get_int();
```
Actually I think this is more correct:
 - raise `RPC_INVALID_PARAMETER` if `node_id < 0`;
 - raise `RPC_INVALID_ADDRESS_OR_KEY` if `!GetNodeStats(...)`.

Note: if you do this then add a test for 1st case.

jnewbery replied:
I don't think it's too important, but feel free to open a follow-up PR if you think it's required.

promag commented on test/functional/net.py:
```diff
@@ -94,5 +95,31 @@ def _test_getpeerinfo(self):
+        assert not peer_info['addnode']
+
+        # Check that updatepeer fails for an invalid peer_id
+        assert_raises_rpc_error(-8, "not found", self.nodes[0].updatepeer, 5000)
```
Missing failure tests for getpeerinfo (maybe unrelated to this PR).

promag reviewed (commented):
Nit, rename PR to *Add updatepeer RPC*. Squash `[RPC] Allow manual_connection to be updated using updatepeer` into `[RPC] Add updatenode RPC`.

jnewbery commented:
@promag nits addressed in latest commit.

promag commented on src/rpc/net.cpp:
```diff
@@ -121,73 +184,82 @@ UniValue getpeerinfo(const JSONRPCRequest& request)
+        throw JSONRPCError(RPC_CLIENT_P2P_DISABLED, "Error: Peer-to-peer functionality missing or disabled");
+    }
+
+    LOCK(cs_main);
```
Remove lock or am I missing something?

jnewbery replied:
you're right - I don't think this is required

promag commented:
utACK after squash.

Nit, how about `updatepeer id setting value (setting value ...)`:
```
updatepeer 1 whitelisted true
updatepeer 1 manual_connection false
```
It works well without named arguments with the cli, and scales well for other settings.

Not, I always prefer named arguments, specially in these calls. But we do support index arguments and this approach doesn't help for future settings.

promag reviewed (commented):
utACK after squash.

Nit, how about `updatepeer id setting value (setting value ...)`:
```
updatepeer 1 whitelisted true
updatepeer 1 manual_connection false
```
It works well without named arguments with the cli, and scales well for other settings.

Note, I always prefer named arguments, specially in these calls. But we do support index arguments and this approach doesn't help for future settings.

NicolasDorier commented:
I agree with @promag `updatepeer id setting value (setting value ...)` would be easier to extend later.

jnewbery commented:
> Nit, how about updatepeer id setting value (setting value ...)

This is a different scheme from all of the existing RPCs. I expect that there would need to be changes to the rpc framework to make this work.

It doesn't make sense to me have a completely different scheme for just this RPC method, since we already have named arguments.

If you disagree, perhaps you could implement the scheme you're talking about in a new branch?

NicolasDorier commented:
@jnewbery how so? This is just 3 differents parameters of 3 strings.

The way you are doing now, if there is like 60 settings for one peer, we will have 60 parameters to this function. This does not seem very maintainable. It would be better in such case to have `updatewhitelist id true`.

promag commented:
> This is designed to be called using named arguments, but due to the RPC infrastructure can also be called with positional arguments (although doing so would be very fiddly).

@jnewbery you too agree that is by design bad for positional arguments (considering the possible settings can be extended).

Another approach is to use something like `sendmany`: `updatepeer {\"setting\":value, ...}`.

FWIW I already gave my utACK 😄

jnewbery commented:
@NicolasDorier 

> how so? This is just 3 differents parameters of 3 strings

Right, but all RPCs currently support positional arguments, where the position of the argument determines its meaning. I don't necessarily agree that it's always appropriate, but having a single RPC that doesn't adhere to that scheme is confusing and inconsistent.

Using named arguments have been around for at least two releases now, so there's no need to use positional arguments at all.

Rebased

NicolasDorier commented:
yeah we talked about it, I understand now why it makes sense named args. Concept ACK.

jnewbery commented:
@NicolasDorier - you've given this a *BIG BIG concept ACK*. Mind reviewing? :slightly_smiling_face:

NicolasDorier commented:
Sure will review and test today.

NicolasDorier commented on src/net.cpp:
```diff
@@ -2571,6 +2583,30 @@ bool CConnman::DisconnectNode(NodeId id)
+bool CConnman::SetManualConnection(NodeId id, bool manual)
+{
+    LOCK(cs_vNodes);
+    for(CNode* pnode : vNodes) {
```
The loop will get repeated all over the place for each field. (Already two times)

I think a better approach would be to lock at `updatepeer` level as well as fetch once for all the right CNode. And change fields directly from there.

NicolasDorier commented on src/rpc/net.cpp:
```diff
@@ -121,73 +184,82 @@ UniValue getpeerinfo(const JSONRPCRequest& request)
+    if (!request.params[1].isNull()) {
+        RPCTypeCheckArgument(request.params[1], UniValue::VBOOL);
+        if (!g_connman->SetWhitelisted(node_id, request.params[1].isTrue())) {
+            throw JSONRPCError(RPC_MISC_ERROR, "Failed to update peer whitelisting");
```
See my above comment, by fetching once the node at the start of this method, you can remove all those `if` as you already know if the CNode with such id already exist.

NicolasDorier commented on src/net.cpp:
```diff
@@ -2539,6 +2539,18 @@ size_t CConnman::GetNodeCount(NumConnections flags)
     return nNum;
 }
 
+bool CConnman::GetNodeStats(NodeId nodeid, CNodeStats& stats)
```
This method would also be removed if fetching CNode was done at the updatepeer level.

NicolasDorier commented:
@jnewbery do you need more help on this PR? I am really interested into seeing it merged.

jnewbery commented on src/net.cpp:
```diff
@@ -2553,6 +2565,30 @@ bool CConnman::DisconnectNode(NodeId id)
+bool CConnman::SetWhitelisted(NodeId id, bool fWhitelisted)
+{
+    LOCK(cs_vNodes);
+    for(CNode* pnode : vNodes) {
```
fixed

jnewbery commented:
@NicolasDorier: Sorry for dropping this - I've rebased on master and squashed all nits.

I'm not sure about your suggestion for locking in updatepeer and fetching the `CNode`. `cs_vNodes` isn't currently locked anywhere outsdie `CConman`, which I think is a good property to maintain.

NicolasDorier commented:
@jnewbery in this case, I would suggest to have a method on CConman `UpdateSetting(str,value)` because there is high ratio of monkey copy pasta code everytimes we will need to add one property.

TheBlueMatt commented:
I'm still with @theuni on this one, not super happy with the idea of changing properties about peers that net_processing/net both consider "constant". Would prefer we add some (undocumented?) options to addnode (or a new RPC) which lets you control the flags of a peer as you create a connection.

jnewbery commented:
@theuni's comment was:

> the idea of adding a command that changes the implied state of a peer via its attributes, rather than performing actions explicitly

I don't understand what's meant by 'performing actions explicitly', but I'm happy to modify this PR if there are concrete suggestions.

> not super happy with the idea of changing properties about peers that net_processing/net both consider "constant"

Can you articulate what makes you not happy? Being able to change the properties of a connection without having to delete/recreate that object is generally very useful, since cycling a connection has many side-effects. As far as I can see, the only stateful impact of starting with `fWhitelisted` rather than updating it later is this call to `AddLocal()` in `BindListenPort()`:

```
    if (addrBind.IsRoutable() && fDiscover && !fWhitelisted)
        AddLocal(addrBind, LOCAL_BIND);
```

I can't see any other reason to be concerned about updating `fWhitelisted` dynamically.

TheBlueMatt commented:
I might be slightly more OK with changing fWhitelisted on a peer after its up, as that may be useful "in the real world" outside of tests, but I'm super not a fan of changing manual_connection. One thing @theuni and I have talked about is being more aggressive about CNode representing a "connection handle" where its potentially allowed to have a few constant members which are general information about the connection, eg whether it was automatically or manually added, the remote address, etc.
One thing you might imagine is inserting a peer into a set/map (eg compact blocks HB mode peers set, the (implied) set of nPreferredDownload peers (which we currently track with fPreferredDownload, but we could actually take it as implication based on const values in CNode)) or otherwise performing (between-message-)stateful behavior on a peer based on such (constant) information about a connection. In each of these cases, having such constant information change out from under you could potentially introduce races/bugs.

> cycling a connection has many side-effects.
I'm not sure what side-effects there are that couldn't be reproduced on a new connection after cycling the connection. It ends up being an unrealistic test if you dont do that anyway. Maybe I'm just missing the motivation of cases that wouldnt be better tested by cycling a connection?

jnewbery commented:
> One thing you might imagine...

You're describing things that _may_ happen in future changes in abstract terms, particularly when you talk about between-message-stateful behaviour.

`m_manual_connection` is certainly a better name than `fAddNode`, but I still don't think it really captures what the property means. Really it's something like 'preferred_peer' - it's a peer that we don't want to disconnect or punish for bad behaviour. How it was connected isn't really relevant. We may want to prefer a peer that has connected inbound to us, and we may want to connect outbound to a peer without preferring it. In that context, I think it makes sense to want to set this property manually after connection, without having to disconnect and go through the version-verack-getheaders-headers handshaking.

This is undoubtedly useful for testing, but I could also imagine it being useful, for example in a business that has more than one bitcoin node and wants to be able to manage their network topology.

Thinking forward a bit more, I think it'd also be beneficial to break out the whitelist behaviour to be more granular, and again be able to update those properties dynamically.

ryanofsky reviewed (commented):
Haven't been following discussion in much detail but if the problem is `fAddNode` having both const and non-const interpretations, maybe it should just be split into two variables: `m_connection_type` that would be const and record origin of connection and `m_preferred_peer` that would be mutable.

kallewoof commented on src/net.cpp:
```diff
@@ -2577,6 +2589,30 @@ bool CConnman::DisconnectNode(NodeId id)
+    }
+    return false;
+}
+
```
Could you maybe make an (optionally inline) method `find_node_by_id` that the above methods use? Looks a bit copy-pastey.

jnewbery replied:
refactored out

kallewoof commented on src/rpc/net.cpp:
```diff
@@ -65,12 +65,80 @@ UniValue ping(const JSONRPCRequest& request)
-            "\nReturns data about each connected network node as a json array of objects.\n"
+            "\nReturns data about connected network nodes as a json array of objects.\n"
+            "\nArguments:\n"
+            "1. \"id\"         (numeric, optional) Only return information about the peer with the specified id.\n"
```
Nit: No period at end (seems to be the norm). Also add this to the first line (`"getpeerinfo \"id\"\n"`)

jnewbery replied:
fixed

kallewoof commented on src/rpc/net.cpp:
```diff
@@ -121,73 +189,80 @@ UniValue getpeerinfo(const JSONRPCRequest& request)
+{
+    if (request.fHelp || request.params.size() < 1 || request.params.size() > 3)
+        throw std::runtime_error(
+            "updatepeer\n"
```
`"updatepeer \"id\" ( \"whitelisted\" \"manual_connection\" )\n"`

kallewoof commented on src/rpc/net.cpp:
```diff
@@ -121,73 +189,80 @@ UniValue getpeerinfo(const JSONRPCRequest& request)
+
+    CNodeStats stats;
+    if (!g_connman->GetNodeStats(node_id, stats)) {
+        throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, strprintf("Node id %u not found", node_id));
```
Nit: `node_id` is an `int`, so `%d` not `%u`.

jnewbery replied:
fixed

kallewoof reviewed (commented):
utACK

Needs rebase.

jnewbery committed:
[net] Add GetNodeByID helper function.

jnewbery commented:
Fixed @kallewoof review comments.

jnewbery commented:
There doesn't seem to much appetite for this. Closing.

NicolasDorier commented:
I liked it :(

MarcoFalke commented:
Marked "up for grabs". (Conversation is locked, so a new pull will need to be opened either way).
